core/drivers/amd/ps_gpio_driver.c

```diff
@@ -0,0 +1,246 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2024-2025, Advanced Micro Devices, Inc. All rights reserved.
+ *
+ */
+
+#include <assert.h>
+#include <drivers/gpio.h>
+#include <gpio_private.h>
+#include <io.h>
+#include <kernel/dt.h>
+#include <kernel/boot.h>
+#include <kernel/interrupt.h>
+#include <kernel/panic.h>
+#include <libfdt.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <sys/queue.h>
+#include <trace.h>
+#include <util.h>
+
+static struct amd_gbank_data ps_bank = {
+	.label = "ps_gpio",
+	.ngpio = 58,
+	.max_bank = PS_BANK_MAX,
+	.bank_min[0] = 0,
+	.bank_max[0] = 25,
+	.bank_min[3] = 26,
+	.bank_max[3] = 57,
+};
+
+/* Standard GPIO Operations */
+static enum gpio_level ps_gpio_get_value(struct gpio_chip *chip,
+					 unsigned int gpio_pin)
+{
+	uint32_t bank = 0;
+	uint32_t pin = 0;
+	struct amd_gpio_info *ps = container_of(chip, struct amd_gpio_info,
+						chip);
+
+	get_bank_and_pin(ps->bdata, gpio_pin, &bank, &pin);
+
+	if ((io_read32(ps->vbase + DATA_RO_OFFSET(bank)) >> pin) & 1)
+		return GPIO_LEVEL_HIGH;
+
+	return GPIO_LEVEL_LOW;
+}
+
+static void ps_gpio_set_value(struct gpio_chip *chip,
+			      unsigned int gpio_pin,
+			      enum gpio_level level)
+{
+	uint32_t bank = 0;
+	uint32_t pin = 0;
+	uint32_t offset = 0;
+	struct amd_gpio_info *ps = container_of(chip, struct amd_gpio_info,
+						chip);
+
+	get_bank_and_pin(ps->bdata, gpio_pin, &bank, &pin);
+
+	if (pin >= GPIO_NUM_MIN && pin < GPIO_NUM_MAX) {
+		offset = DATA_LSW_OFFSET(bank);
+	} else {
+		pin -= GPIO_NUM_MAX;
+		offset = DATA_MSW_OFFSET(bank);
+	}
+
+	/* Explicitly compare the enum value */
+	if (level == GPIO_LEVEL_HIGH)
+		level = 1;
+	else
+		level = 0;
+
+	level = ~BIT32(pin) & (SHIFT_U32(level, pin) | GPIO_UPPER_MASK);
+
+	io_write32(ps->vbase + offset, level);
+}
+
+static enum gpio_dir ps_gpio_get_dir(struct gpio_chip *chip,
+				     unsigned int gpio_pin)
+{
+	uint32_t bank = 0;
+	uint32_t pin = 0;
+	struct amd_gpio_info *ps = container_of(chip, struct amd_gpio_info,
+						chip);
+
+	get_bank_and_pin(ps->bdata, gpio_pin, &bank, &pin);
+
+	if (io_read32(ps->vbase + DIRM_OFFSET(bank)) & BIT(pin))
+		return GPIO_DIR_OUT;
+
+	return GPIO_DIR_IN;
+}
+
+static void ps_gpio_set_dir(struct gpio_chip *chip,
+			    unsigned int gpio_pin,
+			    enum gpio_dir direction)
+{
+	uint32_t bank = 0;
+	uint32_t pin = 0;
+	uint32_t regval = 0;
+	struct amd_gpio_info *ps = container_of(chip, struct amd_gpio_info,
+						chip);
+
+	get_bank_and_pin(ps->bdata, gpio_pin, &bank, &pin);
+
+	if (direction == GPIO_DIR_OUT) {
+		/* set the GPIO pin as output */
+		regval = io_read32(ps->vbase + DIRM_OFFSET(bank));
+		regval |= BIT(pin);
+		io_write32(ps->vbase + DIRM_OFFSET(bank), regval);
```

Comment:
`io_setbits32(ps->vbase + DIRM_OFFSET(bank), BIT(pin)`
Ditto for lines 114/116.

Ditto for line 121/123 (use `io_clrbits32()`)